### ISSUE TITLE
add marshmallow validation

### ISSUE THREAD

**Issue by jgirardet** [falconry/falcon#1335]
Hi,
I use this for my personal use, like jsonschema validation but with the well known marshmallow library. The only trick is that I re-inject validated_data directly to the view via a "data" argument.

If you'are interested, I can do a PR for that.

thanks for falcon
Jimmy

```python

def validate(schema):
    def decorator(func):
        def wrapper(self, req, resp, *args, **kwargs):
            try:
                validated_data = schema.load(req.media)
            except ValidationError as e:
                raise falcon.HTTPBadRequest("validation error", json.dumps(e.messages))

            return func(self, req, resp, data=validated_data.data, *args, **kwargs)

        return wrapper

    return decorator

class Patient(Schema):
    nom = fields.Str(required=True)
    ddn = fields.Date(required=True)


class PatientRessource:
    @validate(Patient())
    def on_post(self, req, resp, data):
       pass
```

**Comment by jgirardet** [comment]
HEre is a more complete version.
Schema is get by differents ways  : class definition, instanciation, decorator call. It  also retrieve the schema  via a callable `def my_callable(request)` .
Custom marshmallow options maybe passed  as kwargs to the decorator.

```python
def validate_marshmallow(custom_schema=None, **schema_kwargs):
    def decorator(func):
        def wrapper(self, req, resp, *args, **kwargs):
            # return empty dict not None
            data = req.media if req.media else {}

            if custom_schema:
                # use either schema or the "schema selector"
                if inspect.isclass(custom_schema) and issubclass(custom_schema, Schema):
                    schema = custom_schema(**schema_kwargs)
                else:
                    schema = custom_schema(req)(**schema_kwargs)

            elif getattr(self, "schema", None):
                # no custom, use instance schema instance or class attribute
                schema = self.schema(**schema_kwargs)

            else:
                # no schema provided, no validation, just return data as it.
                return func(self, req, resp, data=data, *args, **kwargs)

            try:
                validated_data = schema.load(data)
            except ValidationError as e:
                raise falcon.HTTPBadRequest("validation error", e.messages)
            return func(self, req, resp, data=validated_data, *args, **kwargs)

        return wrapper

    return decorator
```

**Comment by GabrielC101** [comment]
Should this be a hook or middleware?

**Comment by vytas7** [comment]
@GabrielC101 it could be either or, Falcon leaves that choice for you:

* A hook applies only to a responder method, or a resource class (the latter may be a bit odd for since you probably do not want to validate ``on_get`` and ``on_post`` the same way, and you probably do not have the same schema for ``on_put`` and ``on_patch``)
* middleware applies to all requests, which would also require some care to check the request method as well as fetch an appropriate schema, maybe from a resource (class?) attribute as an idea

To summarize, the simplest and most general way is a hook, or just a hook-like decorator ( [falcon.media.validators.jsonschema](https://falcon.readthedocs.io/en/stable/_modules/falcon/media/validators/jsonschema.html#validate) is implemented this way), however, for your specific application you may be able to DRY it further by generalizing it with a specialized middleware.
